### PR TITLE
Services (moved Upd, Create & Del into Services)

### PR DESCRIPTION
**Pull request by SasukeVapes**
Created Services, minor changes in controllers (almost all return's have mappings now )

### CODE REVIEW

File: PokemonReviewApi/PokemonReviewApi/Controllers/CountryController.cs [hunk 1]
```diff
@@ -83,11 +77,9 @@ public CountryShortViewModel UpdateCountry([FromQuery] int countryId,  [FromBody
         [ProducesResponseType(404)]
         public CountryEntity DeleteCountry(int countryId)
         {
-            var countryToDelete = _countryRepository.GetCountryById(countryId);
-            _countryRepository.DeleteCountry(countryToDelete);
-            
-            return countryToDelete;
+            _countryService.DeleteCountry(countryId);
+            return null;
         }
```
Comment (Genresal) [inline PokemonReviewApi/PokemonReviewApi/Controllers/CountryController.cs:82]: The same point about null and a blank line below the comment.)

File: PokemonReviewApi/PokemonReviewApi/Controllers/CategoryController.cs [hunk 2]
```diff
@@ -73,9 +68,8 @@ public CategoryShortViewModel UpdateCategory([FromQuery] int categoryId, [FromBo
         [ProducesResponseType(404)]
         public CategoryEntity DeleteCategory(int categoryId)
         {
-            var categoryToDelete = _categoryRepository.GetCategoryById(categoryId);
-            _categoryRepository.DeleteCategory(categoryToDelete);
-            return categoryToDelete;
+            _categoryService.DeleteCategory(categoryId);
+            return null;
         }
```
[the remark this answers is not in the document]
Comment (SasukeVapes) [reply PokemonReviewApi/PokemonReviewApi/Controllers/CategoryController.cs:73]: At some point in the future I might need to return deletedCategory or even deleted categoryId for informational purposes or debugging. Therefore I'd like to keep "return null" for now.

**Review by Genresal — APPROVED** [review]

**Review by GodlyCat — CHANGES_REQUESTED** [review]
Why not for all controllers was created services. 
Why you keep use repositories in controllers
Why not all methods implemented inside services

File: PokemonReviewApi/PokemonReviewApi/Controllers/CategoryController.cs [hunk 3]
```diff
@@ -1,57 +1,49 @@
...
+        public CategoryShortViewModel CreateCategory([FromBody] CategoryShortViewModel categoryCreate)
         {
-            var category = _categoryRepository.GetCategories()
-                .FirstOrDefault(c => c.Name.Trim().ToUpper() == categoryCreate.Name.TrimEnd().ToUpper());
-
-            var categoryMap = _mapper.Map<CategoryEntity>(categoryCreate);
-            _categoryRepository.CreateCategory(categoryMap);
-            return category;
+            var categoryMap= _mapper.Map<Category>(categoryCreate);
+            _categoryService.CreateCategory(categoryMap);
```
Comment (GodlyCat) [inline PokemonReviewApi/PokemonReviewApi/Controllers/CategoryController.cs:45]: Return response from service on create (map model to full vm and return this)

File: PokemonReviewApi/PokemonReviewApi/Controllers/CountryController.cs [hunk 4]
```diff
@@ -1,93 +1,79 @@
...
-
-            var country = _countryRepository.GetCountries()
-                .FirstOrDefault(c => c.Name.Trim().ToUpper() == countryCreate.Name.TrimEnd().ToUpper());
-
-            var countryMap = _mapper.Map<CountryEntity>(countryCreate);
-            _countryRepository.CreateCountry(countryMap);
-
-            return country;
+            var countryMap = _mapper.Map<Country>(countryCreate);
+            _countryService.CreateCountry(countryMap);
```
Comment (GodlyCat) [inline PokemonReviewApi/PokemonReviewApi/Controllers/CountryController.cs:54]: same

File: PokemonReviewApi/PokemonReviewApi/Controllers/CategoryController.cs [hunk 5]
```diff
@@ -60,22 +52,19 @@ public CategoryEntity CreateCategory([FromBody] CategoryShortViewModel categoryC
         [ProducesResponseType(404)]
         public CategoryShortViewModel UpdateCategory([FromQuery] int categoryId, [FromBody] CategoryShortViewModel updatedCategory)
         {
-            var categoryMap = _mapper.Map<CategoryEntity>( updatedCategory);
-            categoryMap.Id = categoryId;
-            _categoryRepository.UpdateCategory(categoryMap);
-           
+            var categoryMap = _mapper.Map<Category>(updatedCategory);
+            _categoryService.UpdateCategory(categoryId, categoryMap);
```
Comment (GodlyCat) [inline PokemonReviewApi/PokemonReviewApi/Controllers/CategoryController.cs:56]: same

File: PokemonReviewApi/PokemonReviewApi/Controllers/CountryController.cs [hunk 6]
```diff
@@ -1,93 +1,79 @@
...
         [ProducesResponseType(404)]
-        public CountryShortViewModel UpdateCountry([FromQuery] int countryId,  [FromBody] CountryShortViewModel updatedCountry)
+        public CountryShortViewModel UpdateCountry([FromQuery] int countryId, [FromBody] CountryShortViewModel updatedCountry)
         {
-            var countryMap = _mapper.Map<CountryEntity>(updatedCountry);
-            countryMap.Id = countryId;
-            _countryRepository.UpdateCountry(countryMap);
-           
+            var countryMap = _mapper.Map<Country>(updatedCountry);
+            _countryService.UpdateCountry(countryId, countryMap);
```
Comment (GodlyCat) [inline PokemonReviewApi/PokemonReviewApi/Controllers/CountryController.cs:65]: same

File: PokemonReviewApi/PokemonReviewApi/Controllers/OwnerController.cs [hunk 7]
```diff
@@ -1,105 +1,84 @@
...
-            var owners = _ownerRepository.GetOwners()
-                .FirstOrDefault(c => c.LastName.Trim().ToUpper() == ownerCreate.LastName.TrimEnd().ToUpper());
-
-            var ownerMap = _mapper.Map<OwnerEntity>(ownerCreate);
-            ownerMap.Country = _countryRepository.GetCountryById(countryId);
-            _ownerRepository.CreateOwner(ownerMap);
-            
-            return owners;
+            var ownerMap = _mapper.Map<Owner>(ownerCreate);
+            _ownerService.CreateOwner(countryId, ownerMap);
```
Comment (GodlyCat) [inline PokemonReviewApi/PokemonReviewApi/Controllers/OwnerController.cs:59]: same

File: PokemonReviewApi/PokemonReviewApi/Controllers/OwnerController.cs [hunk 8]
```diff
@@ -1,105 +1,84 @@
...
         [ProducesResponseType(204)]
         [ProducesResponseType(404)]
-        public OwnerShortViewModel UpdateOwner([FromQuery] int ownerId, [FromBody] OwnerShortViewModel updatedOwner)
+        public OwnerShortViewModel UpdateOwner([FromQuery] int countryId, [FromQuery] int ownerId, [FromBody] OwnerShortViewModel updatedOwner)
         {
-            var ownerMap = _mapper.Map<OwnerEntity>(updatedOwner);
-            ownerMap.Id= ownerId;
-            _ownerRepository.UpdateOwner(ownerMap);
+            var ownerMap = _mapper.Map<Owner>(updatedOwner);
+            _ownerService.UpdateOwner(countryId, ownerId, ownerMap);
```
Comment (GodlyCat) [inline PokemonReviewApi/PokemonReviewApi/Controllers/OwnerController.cs:70]: same

File: PokemonReviewApi/PokemonReviewApi/Controllers/CategoryController.cs [hunk 9]
```diff
@@ -1,57 +1,49 @@
...
-            var categories = _mapper.Map<List<CategoryViewModel>>(_categoryRepository.GetCategories()); //without automapper - var categories = _categoryRepository.GetCategory()
-
-            return categories;
+            return _mapper.Map<List<CategoryViewModel>>(_categoryService.GetCategories());
         }
 
         [HttpGet("{categoryId}")]
         [ProducesResponseType(200, Type = typeof(CategoryEntity))]
         [ProducesResponseType(400)]
         public CategoryShortViewModel GetCategory(int categoryId)
```
Comment (GodlyCat) [inline PokemonReviewApi/PokemonReviewApi/Controllers/CategoryController.cs:34]: why you return short vm ?

**Review by GodlyCat — CHANGES_REQUESTED** [review]
We shouldn't return short vm's from controller 

Same for services, u should return response from repositories on create and update

File: PokemonReviewApi/PokemonReviewApi/Services/IServices/IPokemonService.cs [hunk 10]
```diff
@@ -1,12 +1,19 @@
...
 using PokemonReviewApi.ViewModels;
 
 namespace PokemonReviewApi.Services.IServices
 {
     public interface IPokemonService
     {
-        PokemonEntity CreatePokemon(int ownerId, int categoryId, PokemonEntity createdPokemon);
-        PokemonShortViewModel UpdatePokemon(int pokeId, int ownerId, int categoryId, PokemonShortViewModel updatedPokemon);
-        void DeletePokemon(PokemonEntity deletedPokemon);
+        List<Pokemon> GetPokemons();
```
Comment (Glipchik) [inline PokemonReviewApi/PokemonReviewApi/Services/IServices/IPokemonService.cs:8]: You probably could add blank lines between these methods to improve the visual perception

File: PokemonReviewApi/PokemonReviewApi/Helper/MappingProfiles.cs [hunk 11]
```diff
@@ -1,27 +1,21 @@
...
-
-            CreateMap<OwnerEntity, OwnerViewModel>().ReverseMap();
-            CreateMap<OwnerShortViewModel, OwnerEntity>().ReverseMap();
-
-            CreateMap<PokemonEntity, PokemonViewModel>().ReverseMap();
-            CreateMap<PokemonShortViewModel, PokemonEntity>().ReverseMap();
-           
-            
-            
+            CreateMap<Category, CategoryViewModel>().ReverseMap();
```
Comment (Glipchik) [inline PokemonReviewApi/PokemonReviewApi/Helper/MappingProfiles.cs:11]: And some blank lines between mappings of different entities might be useful here too

File: PokemonReviewApi/PokemonReviewApi/Services/IServices/IPokemonService.cs (hunk 10 above)
Comment (Glipchik) [reply PokemonReviewApi/PokemonReviewApi/Services/IServices/IPokemonService.cs:8]: Also it is not the best idea to put the folder with interfaces inside the folder with services. You may create the separate folder in the project called 'Interfaces' or 'Abstractions' and put these interfaces there

**Review by Glipchik — APPROVED** [review]

**Review by GodlyCat — APPROVED** [review]

**Pull request merged**